pip/pip-392.md

```diff
@@ -0,0 +1,94 @@
+# PIP-392: Add configuration to enable consistent hashing to select active consumer for partitioned topic
+
+# Background knowledge
+
+After [#19502](https://github.com/apache/pulsar/pull/19502) will use consistent hashing to select active consumer for non-partitioned topic
+
+# Motivation
+
+Currently, for partitioned topics, the active consumer is selected using the formula [partitionedIndex % consumerSize](https://github.com/apache/pulsar/blob/137df29f85798b00de75460a1acb91c7bc25453f/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractDispatcherSingleActiveConsumer.java#L129-L130). 
+This method can lead to uneven distribution of active consumers.
+
+Consider a scenario with 100 topics named `public/default/topic-{0~100}`, each having `one partition`. 
+If 10 consumers are created using a `regex` subscription with the `Failover type`, all topic will be assigned to the same consumer(the first connected consumer). This results in an imbalanced distribution of consumers.
+
+# Goals
+
+## In Scope
+- Address the issue of imbalance for `failover` subscription type consumers in single-partition or few-partition topics.
```

Comment:
This change will introduce a new problem: Let's say that there's a single connected consumer and another consumer connects, it could switch to use the new consumer. When this happens, there's an existing problem with failover and exclusive consumers where the outstanding messages aren't properly drained before switching. 
There's a stalled PIP, [PIP-260: Client consumer filter received messages](https://github.com/apache/pulsar/issues/19864), which is related to the issue, however it's not exactly describing the consumer switching scenario. For non-partitioned topics there's a reported issue #15189. After PIP-392 changes, a similar problem would also appear for partitioned topics.
I guess addressing #15189 could happen in a separate PIP, but I'd suggest documenting this detail also in the PIP-392 document that it's a known consequence of the change.
For consumer switching in failover subscriptions, there should be a way to ensure that pending messages are processed before switching the consumer. Otherwise it leads to duplicate message processing in clients which could be considered as out-of-order message processing from the client application perspective.

Reply:
hi, @lhotari Thanks for bringing up this topic. 

Here's my conclusion: **Pulsar current cannot avoid duplicate messages on the client side during broker or active consumer transfer scenarios, and this PIP doesn't introduce new issues because both consistent hashing and modulo algorithms will cause active consumer transfer.**

Please help review my explanation.


### For broker ownership transfer case: https://github.com/apache/pulsar/issues/19864
This is a known case, so let's not discuss it too much here.


### For active consumer transfer case(Failover subscription)

#### 1. Use consistent hashing algorithm

Before this PIP, we only used the consistent hashing algorithm with `non-partition` topic.

Since the consistent hashing algorithm creates a hash ring based on the consumer's name, when a new consumer joins, new nodes are added to the hash ring, causing topics to move to this consumer. Thus, topics are transferred.

Refer to the image: The `topic-3` transfers from `c2` to `c4`.

<img width="1677" alt="image" src="https://github.com/user-attachments/assets/07f50743-f573-49e3-8e4a-f0570247b485">

This issue exists with non-partitioned topics because they use consistent hashing. I agree with your point that after this PIP, it will also occur with partitioned topics.

However, I want to emphasize that this issue can occur even without consistent hashing for `partitioned topic`. Please see the explanation below.


#### 2. Use mod(partitionedIndex % consumerSize)   - **current partitioned topic algorithm**

Assume a topic has 2 partition: **partitionedIndex [P0, P1]**

1. First consumer added(**consumer-0**)
  - **P0** belong **consumer-0**: (paritonIndex:0)%(consumerSize:1)=0, 
  - **P1** belong  **consumer-0**: (paritonIndex:1)%(consumerSize:1)=0, 
2. When secnd consumer is added(consumer-1) 
  - **P0** belong **consumer-0**: (paritonIndex:0)%(consumerSize:2)=0, 
  - **P1** belong  **consumer-1**: (paritonIndex:1)%(consumerSize:2)=1, 

P1 activeConsumer from consumer-0 transfer to consumer-1.

P1's active consumer transfers from consumer-0 to consumer-1.

Therefore, P1 undergoes an active consumer transfer.

### So go back this comment
> After PIP-392 changes, a similar problem would also appear for partitioned topics.
I guess addressing https://github.com/apache/pulsar/issues/15189 could happen in a separate PIP, but I'd suggest documenting this detail also in the PIP-392 document that it's a known consequence of the change.

I don't think it's necessary to add more explanation for this.

1. #15189 does not have a clear conclusion proving that the consistent hash algorithm for non-partitioned topics has issues.
2. Regarding potential duplicate messages and out-of-order behavior in failover, we have already explained this in the documentation: https://pulsar.apache.org/docs/4.0.x/concepts-messaging/#failover"

Reply:
> Here's my conclusion: **Pulsar current cannot avoid duplicate messages on the client side during broker or active consumer transfer scenarios, and this PIP doesn't introduce new issues because both consistent hashing and modulo algorithms will cause active consumer transfer.**

The way how this PR introduces more frequent problems is that the consumer will switch frequently when new consumers get added. A rolling restart is a common problematic case. This is why the PIP document should at least document that this is a known problem and consequence of this change.

I hope that @codelipenghui and @merlimat could share their view on this matter.

Reply:
> The way how this PR introduces more frequent problems is that the consumer will switch frequently when new consumers get added. A rolling restart is a common problematic case. This is why the PIP document should at least document that this is a known problem and consequence of this change.

If I understand correctly, the modulo algorithm also has this situation.

Assuming there are 5 partitions with indices partitionIndex: [0, 1, 2, 3, 4]

#### 1. The first consumer-0 joins, partitions selected:
  - consumer-0: [0, 1, 2, 3, 4]
```
partitionIndex % consumerSize = consumerIndex
0 % 1 = 0
1 % 1 = 0
2 % 1 = 0
3 % 1 = 0
4 % 1 = 0
```

#### 2. When the second consumer-1 joins, partitions selected:
   - consumer-0: [0, 2, 4]  
   - consumer-1: [1, 3]
```
partitionIndex % consumerSize = consumerIndex
0 % 2 = 0
1 % 2 = 1
2 % 2 = 0
3 % 2 = 1
4 % 2 = 0
```
`!!`**!!P1 and P3** also get transferred

#### 3. When the third consumer-2 joins, partitions selected:
   - consumer-0: [0, 3]  
   - consumer-1: [1, 4]
   - consumer-2: [2]  
```
partitionIndex % consumerSize = consumerIndex
0 % 3 = 0
1 % 3 = 1
2 % 3 = 2
3 % 3 = 0
4 % 3 = 1
```
`!!`**P2, P3 and P4** also get transferred

Altogether transferred

1. P1 from c0 -> c1
2. P2 from c0 -> c2
3. P3 from c0 -> c1 -> c2
4. P4 from c0 -> c2

Reply:
@shibd Thanks for clarifying. Makes sense. PIP-392 won't make the problem worse in that case. One of my assumptions was that in the current solution there's "stickyness" in a way that the active consumer doesn't get switched unless it fails. I guess that's a false assumption from my side. 

Although it's out-of-scope for PIP-392, it would be great to start making progress in addressing the clear problem with Failover subscriptions which is documented as "In some cases, a partition may have an older active consumer processing messages while a newly switched over active consumer starts receiving new messages. This may lead to message duplication or out-of-order.". That goes against a lot of the expected guarantees for Failover subscriptions and causes problems. I found out that @grssam provided some feedback about this in the past in https://github.com/apache/pulsar-site/pull/633#issuecomment-1625416461 . It could be valuable if @grssam could also review PIP-392.

Reply:
hi, @lhotari I added this to `Out of Scope` section. PTAL.

Reply:
Thanks, this resolves this discussion.

Reply:
yea, i don't see hash based having a bigger impact on the rebalance that happens during things like consumer pod rollout etc. maybe the situation would actually be better as previously it was skewed (for multiple partitioned topics in a single sub) so impact was higher, but now based on how evenly things are spread, not the same amount of partitions would get rebalanced.